src/lib/trakeva_cache.ml

```diff
@@ -0,0 +1,195 @@
+(**************************************************************************)
+(*  Copyright 2014, Sebastien Mondet <seb@mondet.org>                     *)
+(*                                                                        *)
+(*  Licensed under the Apache License, Version 2.0 (the "License");       *)
+(*  you may not use this file except in compliance with the License.      *)
+(*  You may obtain a copy of the License at                               *)
+(*                                                                        *)
+(*      http://www.apache.org/licenses/LICENSE-2.0                        *)
+(*                                                                        *)
+(*  Unless required by applicable law or agreed to in writing, software   *)
+(*  distributed under the License is distributed on an "AS IS" BASIS,     *)
+(*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or       *)
+(*  implied.  See the License for the specific language governing         *)
+(*  permissions and limitations under the License.                        *)
+(**************************************************************************)
+
+open Nonstd
+open Pvem_lwt_unix
+open Pvem_lwt_unix.Deferred_result
+module String = StringLabels
+
+let debug = ref false
+
+let dbg fmt = ksprintf (eprintf "Trakeva_cache: %s\n%!") fmt
+    
+module Greedy_cache = struct
+
+  type t = {
+    collections: (string, (string, string) Hashtbl.t) Hashtbl.t;
+  }
+  let create () = {collections = Hashtbl.create 10}
+  let find_collection {collections} c =
+    try Some (Hashtbl.find collections c)
+    with _ -> None
+
+  let from_collection c key =
+    try Some (Hashtbl.find c key)
+    with _ -> None
+    
+  let new_collection t collection =
+    let col = Hashtbl.create 42 in
+    Hashtbl.add t.collections collection col;
+    col
+
+  let add_to_collection t col ~key ~value =
+    Hashtbl.add col key value
+
+  let fold_collection col ~init ~f =
+    let r = ref init in
+    Hashtbl.iter (fun key value ->
+        if !debug then dbg "fold_collection %s, %s" key value;
+        r := f !r ~key ~value) col;
+    !r
+
+
```

Comment:
These names (`from_collection`, `new_collection`, `add_to_collection` and `fold_collection`) seem redundant. Why not rename the module and just have `find`, `new`, `add`, and `fold` ?